File: src/core/predicate.zig
const std = @import("std");

pub const Operator = enum {
    eq,
    neq,
    gt,
    lt,
    contains,
    in_list,
};

pub const Condition = struct {
    column: []const u8,
    op: Operator,
    value: []const u8,
};

pub const ResolvedCondition = struct {
    index: usize,
    op: Operator,
    value: []const u8,
};

pub fn eq(lhs: []const u8, rhs: []const u8) bool {
    return std.mem.eql(u8, lhs, rhs);
}

pub fn parseConditions(allocator: std.mem.Allocator, text: []const u8) ![]Condition {
    var out: std.ArrayList(Condition) = .empty;
    defer out.deinit(allocator);

    var parts = std.mem.splitScalar(u8, text, ',');
    while (parts.next()) |part| {
        if (part.len == 0) continue;
        const p = std.mem.trim(u8, part, " ");
        if (p.len == 0) continue;

        if (std.mem.indexOf(u8, p, " contains ")) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + " contains ".len ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .contains, .value = val });
            continue;
        }
        if (std.mem.indexOf(u8, p, " in ")) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + " in ".len ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .in_list, .value = val });
            continue;
        }
        if (std.mem.indexOf(u8, p, "!=")) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + 2 ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .neq, .value = val });
            continue;
        }
        if (std.mem.indexOfScalar(u8, p, '>')) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + 1 ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .gt, .value = val });
            continue;
        }
        if (std.mem.indexOfScalar(u8, p, '<')) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + 1 ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .lt, .value = val });
            continue;
        }
        if (std.mem.indexOfScalar(u8, p, '=')) |sep| {
            const col = std.mem.trim(u8, p[0..sep], " ");
            const val = std.mem.trim(u8, p[sep + 1 ..], " ");
            if (col.len == 0) return error.InvalidFilterExpression;
            try out.append(allocator, .{ .column = col, .op = .eq, .value = val });
            continue;
        }

        return error.InvalidFilterExpression;
    }
    return out.toOwnedSlice(allocator);
}

pub fn rowMatchesAll(fields: []const []const u8, conditions: []const ResolvedCondition) bool {
    for (conditions) |cond| {
        if (cond.index >= fields.len) return false;
        if (!matchField(fields[cond.index], cond)) return false;
    }
    return true;
}

fn matchField(field: []const u8, cond: ResolvedCondition) bool {
    return switch (cond.op) {
        .eq => eq(field, cond.value),
        .neq => !eq(field, cond.value),
        .contains => std.mem.indexOf(u8, field, cond.value) != null,
        .gt => compareNumber(field, cond.value, .gt),
        .lt => compareNumber(field, cond.value, .lt),
        .in_list => inList(field, cond.value),
    };
}

const Cmp = enum { gt, lt };

fn compareNumber(field: []const u8, rhs_text: []const u8, cmp: Cmp) bool {
    const lhs = std.fmt.parseFloat(f64, field) catch return false;
    const rhs = std.fmt.parseFloat(f64, rhs_text) catch return false;
    return switch (cmp) {
        .gt => lhs > rhs,
        .lt => lhs < rhs,
    };
}

fn inList(field: []const u8, raw_list: []const u8) bool {
    var items = std.mem.splitScalar(u8, raw_list, '|');
    while (items.next()) |item| {
        if (eq(field, std.mem.trim(u8, item, " "))) return true;
    }
    return false;
}

File: csv-utils-core/src/predicate.rs
use crate::schema;
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Operator {
    Eq,
    Neq,
    Gt,
    Lt,
    Contains,
    InList,
}

#[derive(Debug, Clone)]
pub struct Condition {
    pub column: String,
    pub op: Operator,
    pub value: String,
}

#[derive(Debug, Clone)]
pub struct ResolvedCondition {
    pub index: usize,
    pub op: Operator,
    pub value: String,
}

#[derive(Debug, thiserror::Error)]
pub enum PredicateError {
    #[error("invalid filter expression")]
    InvalidExpression,
    #[error("column not found: {0}")]
    ColumnNotFound(String),
}

pub fn parse_conditions(text: &str) -> Result<Vec<Condition>, PredicateError> {
    let mut out = Vec::new();
    for part in text.split(',') {
        let p = part.trim();
        if p.is_empty() {
            continue;
        }

        if let Some(sep) = p.find(" contains ") {
            let col = p[..sep].trim();
            let val = p[sep + " contains ".len()..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::Contains,
                value: val.to_string(),
            });
            continue;
        }
        if let Some(sep) = p.find(" in ") {
            let col = p[..sep].trim();
            let val = p[sep + " in ".len()..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::InList,
                value: val.to_string(),
            });
            continue;
        }
        if let Some(sep) = p.find("!=") {
            let col = p[..sep].trim();
            let val = p[sep + 2..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::Neq,
                value: val.to_string(),
            });
            continue;
        }
        if let Some(sep) = p.find('>') {
            let col = p[..sep].trim();
            let val = p[sep + 1..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::Gt,
                value: val.to_string(),
            });
            continue;
        }
        if let Some(sep) = p.find('<') {
            let col = p[..sep].trim();
            let val = p[sep + 1..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::Lt,
                value: val.to_string(),
            });
            continue;
        }
        if let Some(sep) = p.find('=') {
            let col = p[..sep].trim();
            let val = p[sep + 1..].trim();
            if col.is_empty() {
                return Err(PredicateError::InvalidExpression);
            }
            out.push(Condition {
                column: col.to_string(),
                op: Operator::Eq,
                value: val.to_string(),
            });
            continue;
        }

        return Err(PredicateError::InvalidExpression);
    }
    Ok(out)
}

pub fn resolve_conditions(
    headers: &[String],
    conditions: &[Condition],
) -> Result<Vec<ResolvedCondition>, PredicateError> {
    let mut resolved = Vec::with_capacity(conditions.len());
    for cond in conditions {
        let index = schema::index_of(headers, &cond.column)
            .ok_or_else(|| PredicateError::ColumnNotFound(cond.column.clone()))?;
        resolved.push(ResolvedCondition {
            index,
            op: cond.op,
            value: cond.value.clone(),
        });
    }
    Ok(resolved)
}

pub fn row_matches_all(fields: &[String], conditions: &[ResolvedCondition]) -> bool {
    conditions.iter().all(|cond| {
        fields
            .get(cond.index)
            .is_some_and(|field| match_field(field, cond))
    })
}

fn match_field(field: &str, cond: &ResolvedCondition) -> bool {
    match cond.op {
        Operator::Eq => field == cond.value,
        Operator::Neq => field != cond.value,
        Operator::Contains => field.contains(&cond.value),
        Operator::Gt => compare_number(field, &cond.value, |a, b| a > b),
        Operator::Lt => compare_number(field, &cond.value, |a, b| a < b),
        Operator::InList => cond
            .value
            .split('|')
            .map(str::trim)
            .any(|item| field == item),
    }
}

fn compare_number(field: &str, rhs_text: &str, cmp: fn(f64, f64) -> bool) -> bool {
    let Ok(lhs) = field.parse::<f64>() else {
        return false;
    };
    let Ok(rhs) = rhs_text.parse::<f64>() else {
        return false;
    };
    cmp(lhs, rhs)
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Operator::Eq => write!(f, "="),
            Operator::Neq => write!(f, "!="),
            Operator::Gt => write!(f, ">"),
            Operator::Lt => write!(f, "<"),
            Operator::Contains => write!(f, "contains"),
            Operator::InList => write!(f, "in"),
        }
    }
}
